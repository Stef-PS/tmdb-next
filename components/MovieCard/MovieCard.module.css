.card {
  justify-self: start;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  position: relative;
  cursor: pointer;
  padding: var(--spacing-xs);
}

.poster {
  display: block;
  width: 100%;
  box-shadow: var(--elevation-mid);
}

.card:hover .poster {
  box-shadow: var(--elevation-high);
}

.title {
  font-size: var(--font-size-xl);
  font-weight: var(--font-weight-bold);
  margin: var(--spacing-xs) 0;
  text-align: center;
  flex: 1;
}

.info {
  width: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
}

.year {
  font-weight: var(--font-weight-semibold);
}

.aside {
  padding: var(--spacing-xs);
  display: flex;
  flex-direction: column;
  text-align: center;
}