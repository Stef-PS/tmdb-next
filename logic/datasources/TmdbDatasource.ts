import axios from 'axios'

const LOCALE = 'en-US'
const API_URL = 'https://api.themoviedb.org/3'

export interface SearchMovie {
  id: number,
  poster_path: string,
  adult: boolean,
  overview: string,
  release_date: string,
  genre_ids: number[],
  original_title: string,
  original_language: string,
  title: string,
  backdrop_path: string,
  popularity: number,
  vote_count: number,
  video: boolean,
  vote_average: number,
}

export interface SearchMovieResult {
  page: number,
  results: SearchMovie[],
  total_results: number,
  total_pages: number,
}

export interface Configuration {
  images: {
    baseUrl: string,
    secureBaseUrl: string,
    posterSizes: string[]
  }
}

export class TmdbDatasource {
  private apiKey = process.env.TMDB_API_KEY ?? 'test'
  private options = {
    timeout: 2500,
    headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' }
  }

  searchMovie(query: string, page = 1): Promise<any> {
    const url = `${API_URL}/search/movie?api_key=${this.apiKey}&query=${query}&page=${page}&language=${LOCALE}`
    return axios.get(url, this.options)
      .then((response) => response.data)
      .catch(error => {
        return Promise.reject(error)
      })
  }

  getConfiguration(): Promise<Configuration> {
    const url = `${API_URL}/configuration?api_key=${this.apiKey}`
    return axios.get(url, this.options)
      .then((response) => {
        const {
          base_url: baseUrl,
          secure_base_url: secureBaseUrl,
          poster_sizes: posterSizes
        } = response.data.images
        return { images: { baseUrl, secureBaseUrl, posterSizes } }
      })
      .catch(error => {
        return Promise.reject(error)
      })
  }
}
