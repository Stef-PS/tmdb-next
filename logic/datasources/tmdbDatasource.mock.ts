export const searchMovieResultMock = (query: string, page: number, pageSize: number, total: number) => {
  const results = []
  for (let i = pageSize * (page - 1); i < Math.min(pageSize * page, total); i++) {
    results.push({
      id: i,
      poster_path: `poster_path_${i}`,
      adult: false,
      overview: `overview_${i}`,
      release_date: 'release_date',
      genre_ids: [1, 2],
      original_title: `original_title_${i}`,
      original_language: 'original_language',
      title: `title_${query}_${i}`,
      backdrop_path: `backdrop_path_${i}`,
      popularity: 1,
      vote_count: 1,
      video: false,
      vote_average: 1,
    })
  }
  return {
    page: page,
    results,
    total_results: total,
    total_pages: page,
  }
}

export const getConfigurationMock = () => {
  return {
    images: {
      base_url: 'http://image.tmdb.org',
      secure_base_url: 'https://image.tmdb.org',
      poster_sizes: ['w500']
    }
  }
}

export const getMovieMock = (id: number) => ({
  adult: false,
  backdrop_path: `backdrop_path_${id}`,
  belongs_to_collection: {
    id: id,
    name: `name_${id}`,
    poster_path: `poster_path_${id}`,
    backdrop_path: `backdrop_path_${id}`
  },
  budget: id,
  genres: [{ id: id, name: `name_${id}` }],
  homepage: `homepage_${id}`,
  id: id,
  imdb_id: `imdb_id_${id}`,
  original_language: 'original_language',
  original_title: `original_title_${id}`,
  overview: `overview_${id}`,
  popularity: id,
  poster_path: `poster_path_${id}`,
  production_companies: [{ id: id, logo_path: `logo_path_${id}`, name: `name_${id}`, origin_country: `origin_country_${id}` }],
  production_countries: [{ iso_3166_1: `iso_3166_1_${id}`, name: `name_${id}` }],
  release_date: 'release_date',
  revenue: id,
  runtime: id,
  spoken_languages: [{ english_name: `english_name_${id}`, iso_639_1: `iso_639_1_${id}`, name: `name_${id}` }],
  status: 'status',
  tagline: `tagline_${id}`,
  title: `title_${id}`,
  video: false,
  vote_average: id,
  vote_count: id
})
