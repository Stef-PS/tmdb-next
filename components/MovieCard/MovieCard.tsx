import Image from "next/image"
import Rating from '../Rating'
import { Configuration, SearchMovie } from "../../logic/datasources/TmdbDatasource"
import styles from './MovieCard.module.css'

interface MovieCardProps {
  movie: SearchMovie,
  config: Configuration
}

const MovieCard = ({ movie, config }: MovieCardProps) => {
  const image = { width: 342, height: 513 }
  const posterUrl = `${config.images.secureBaseUrl}w${image.width}${movie.posterPath}`

  return (
    <li className={styles.card}>
      <div className={styles.poster}>
        <Image src={posterUrl} alt={movie.title} layout="responsive" width={image.width} height={image.height} />
      </div>
      <div className={styles.info}>
        <div className={styles.title}>{movie.title}</div>
        <div className={styles.aside}>
          <div className={styles.year}>{movie.releaseDate.substring(0, 4)}</div>
          <Rating rate={movie.voteAverage} />
        </div>
      </div>
    </li>
  )
}

export default MovieCard
