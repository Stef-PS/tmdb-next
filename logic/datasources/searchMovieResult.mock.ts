import { SearchMovie, SearchMovieResult } from './TmdbDatasource'

export const searchMovieResultMock = (query: string, page: number, pageSize: number, total: number): SearchMovieResult => {
  const results: SearchMovie[] = []
  for (let i = pageSize * (page - 1); i < Math.min(pageSize * page, total); i++) {
    results.push({
      id: i,
      poster_path: `poster_path_${i}`,
      adult: false,
      overview: `overview_${i}`,
      release_date: 'release_date',
      genre_ids: [1, 2],
      original_title: `original_title_${i}`,
      original_language: 'original_language',
      title: `title_${query}_${i}`,
      backdrop_path: `backdrop_path_${i}`,
      popularity: 1,
      vote_count: 1,
      video: false,
      vote_average: 1,
    })
  }
  return {
    page: page,
    results,
    total_results: total,
    total_pages: page,
  }
}
