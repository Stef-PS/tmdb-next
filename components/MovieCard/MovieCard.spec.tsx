import { render } from '@testing-library/react'
import { Configuration, SearchMovie } from '../../logic/datasources/TmdbDatasource'
import MovieCard from './MovieCard'

const movie: SearchMovie = {
  id: 1,
  title: 'title',
  releaseDate: '2009-09-20',
  voteAverage: 10,
  posterPath: '/poster_path.jpg',
  overview: 'this is a test movie.',
  originalTitle: 'original title',
  originalLanguage: 'original language',
  popularity: 1,
  voteCount: 2,
}

const config: Configuration = {
  images: {
    secureBaseUrl: 'https://image.tmdb.org/t/p/',
    baseUrl: 'https://image.tmdb.org/t/p/',
    posterSizes: ['w342']
  }
}

describe('MovieCard', () => {
  it('should render MovieCard unchanged', () => {
    const { container } = render(<MovieCard movie={movie} config={config} />)
    expect(container).toMatchSnapshot()
  })
})

