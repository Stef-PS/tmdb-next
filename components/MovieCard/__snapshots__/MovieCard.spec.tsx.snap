// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should render MovieCard unchanged 1`] = `
<div>
  <li
    class="card"
  >
    <div
      class="poster"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 150% 0px 0px 0px;"
        />
        <img
          alt="title"
          data-nimg="responsive"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="info"
    >
      <div
        class="title"
      >
        title
      </div>
      <div
        class="aside"
      >
        <div
          class="year"
        >
          2009
        </div>
        <div
          class="rating"
        >
          <span
            class="star"
          >
            ★
          </span>
          <span
            class="star"
          >
            ★
          </span>
          <span
            class="star"
          >
            ★
          </span>
          <span
            class="star"
          >
            ★
          </span>
          <span
            class="star"
          >
            ★
          </span>
        </div>
      </div>
    </div>
  </li>
</div>
`;
